tempodb/blocklist/poller.go

```diff
@@ -226,69 +254,164 @@ func (p *Poller) pollTenantAndCreateIndex(ctx context.Context, tenantID string)
+
+	assembleSpan, _ := opentracing.StartSpanFromContext(derivedCtx, "assemblePreviousResults")
+	mm := make(map[uuid.UUID]*backend.BlockMeta, len(metas))
+	cm := make(map[uuid.UUID]*backend.CompactedBlockMeta, len(compactedMetas))
+
+	for _, i := range metas {
+		mm[i.BlockID] = i
+	}
+
+	for _, i := range compactedMetas {
+		cm[i.BlockID] = i
+	}
+	assembleSpan.Finish()
+
+	mtx := sync.Mutex{}
+	newBlockList := make([]*backend.BlockMeta, 0, len(currentBlockIDs))
+	newCompactedBlocklist := make([]*backend.CompactedBlockMeta, 0, len(currentCompactedBlockIDs))
+
+	learnSpan, _ := opentracing.StartSpanFromContext(derivedCtx, "learnFromPrevious")
+	newBlockIDs := []uuid.UUID{}
+	for _, blockID := range currentBlockIDs {
+		// if we already have this block id in our previous list, use the existing data.
+		if v, ok := mm[blockID]; ok {
+			newBlockList = append(newBlockList, v)
+			continue
+		}
+		newBlockIDs = append(newBlockIDs, blockID)
+	}
 
-	for _, blockID := range blockIDs {
+	for _, blockID := range currentCompactedBlockIDs {
+		// if we already have this block id in our previous list, use the existing data.
+		if v, ok := cm[blockID]; ok {
+			newCompactedBlocklist = append(newCompactedBlocklist, v)
+			continue
+		}
+
+		// TODO: Review this hack to avoid polling for compacted blocks that we
+		// know about.  But we need to poll this to ensure that we have the correct
+		// time. Probably.
+		// if v, ok := mm[blockID]; ok {
+		// 	chCompactedMeta <- &backend.CompactedBlockMeta{
+		// 		BlockMeta: *v,
+		// 		// CompactedTime: time.Now(),
+		// 	}
+		// 	continue
+		// }
+
+		newBlockIDs = append(newBlockIDs, blockID)
+
+		// TODO: for the compacted block IDs that we don't know about, we should
+		// poll directly rather than go through the pollBlock() function.  This
+		// would avoid the 404 on the meta.json and go directly to the
+		// meta.compacted.json.
+		// compactedBlockMeta, err = p.compactor.CompactedBlockMeta(blockID, tenantID)
+		// if err != nil {
+		// 	// log
+		// 	level.Error(p.logger).Log("msg", "failed to poll compacted block", "tenant", tenantID, "blockID", blockID, "err", err)
+		// 	continue
+		// }
+		// if compactedBlockMeta != nil {
+		// 	newCompactedBlocklist = append(newCompactedBlocklist, compactedBlockMeta)
+		// }
+
+	}
+	learnSpan.Finish()
+
+	var errs []error
+
+	pollSpan, pollSpanCtx := opentracing.StartSpanFromContext(derivedCtx, "pollUnknown")
+	pollSpan.SetTag("newBlockIDs", len(newBlockIDs))
+	bg := boundedwaitgroup.New(p.cfg.PollConcurrency)
+	for _, blockID := range newBlockIDs {
 		bg.Add(1)
-		go func(uuid uuid.UUID) {
+
+		mtx.Lock()
+		if len(errs) > 0 {
+			mtx.Unlock()
+			break
+		}
+		mtx.Unlock()
+
+		go func(id uuid.UUID) {
 			defer bg.Done()
 
 			if p.cfg.PollJitterMs > 0 {
 				time.Sleep(time.Duration(rand.Intn(p.cfg.PollJitterMs)) * time.Millisecond)
 			}
 
-			m, cm, err := p.pollBlock(ctx, tenantID, uuid)
+			m, cm, pollBlockErr := p.pollBlock(pollSpanCtx, tenantID, id)
```

Comment:
This call will make 2 reads for every compactedBlock, since it first tries meta.json.  However we already knew the block is compacted from the ListBlocks call.  I think this could be addressed by keeping the `newBlockIDs` slice split between compacted and regular, and then pass that info to pollBlock (maybe an option like `compacted bool`.  So on a compacted block it doesn't need to try meta.json.   I think for a regular block it still has to try both because there is a non-zero chance the block got compacted in between ListBlocks and pollBlock.